lib/Conversion/FIRRTLToHW/LowerToHW.cpp

```diff
@@ -1468,6 +1476,78 @@ Value FIRRTLLowering::getLoweredValue(Value value) {
   return result;
 }
 
+/// Return the lowered array value whose type is converted into `destType`.
+/// We have to care about the extension/truncation/signedness of each element.
+/// If returns a null value for complex arrays such as arrays with bundles.
+Value FIRRTLLowering::getExtOrTruncArrayValue(Value array,
+                                              FIRRTLType sourceType,
+                                              FIRRTLType destType,
+                                              bool allowTruncate) {
+  SmallVector<Value> resultBuffer;
+  bool success = true;
+
+  // Helper function to cast each element of array to dest type.
+  auto cast = [&](Value value, FIRRTLType sourceType, FIRRTLType destType) {
+    auto srcWidth = sourceType.cast<IntType>().getWidthOrSentinel();
+    auto destWidth = destType.cast<IntType>().getWidthOrSentinel();
+    auto resultType = builder.getIntegerType(destWidth);
+
+    if (srcWidth == destWidth)
+      return value;
+
+    if (srcWidth > destWidth) {
+      if (allowTruncate)
+        return builder.createOrFold<comb::ExtractOp>(resultType, value, 0);
+
+      builder.emitError("operand should not be a truncation");
```

Comment:
Where will this error message show up? Is it associated with the operation?

Reply:
This error is emitted to the operation we are lowering, connect op in this case. For example,
```
circuit connectNarraow: 
  module connectNarraow:
    input clock: Clock
    input c: UInt<2>[2]
    output b: UInt<1>[2]
    reg r1: UInt<1>[2], clock
    r1 <= c
    b <= r1
```
for above
```
...
./build_debug/foo.fir:7:8: error: operand should not be a truncation
    r1 <= c
```